engine/runtime/src/main/scala/org/enso/interpreter/runtime/util/DiagnosticFormatter.scala

```diff
@@ -25,17 +26,46 @@ class DiagnosticFormatter(
   private val maxSourceLinesToPrint          = 3
   private val linePrefixSize                 = blankLinePrefix.length
   private val outSupportsAnsiColors: Boolean = outSupportsColors
-  private val (textAttrs: fansi.Attrs, subject: String) = diagnostic match {
-    case _: Error   => (fansi.Color.Red ++ fansi.Bold.On, "error: ")
-    case _: Warning => (fansi.Color.Yellow ++ fansi.Bold.On, "warning: ")
+
+  sealed protected trait DiagnosticKind
+
+  protected object DiagnosticKind {
+    case object Error   extends DiagnosticKind
+    case object Warning extends DiagnosticKind
+  }
+
+  protected val diagnosticKind: DiagnosticKind = diagnostic match {
+    case _: Error   => DiagnosticKind.Error
+    case _: Warning => DiagnosticKind.Warning
     case _          => throw new IllegalStateException("Unexpected diagnostic type")
   }
-  private lazy val both = textAndLocation()
 
-  def format() = both._1
-  def where()  = both._2
+  private val textAttrs: fansi.Attrs = diagnosticKind match {
+    case DiagnosticKind.Error   => fansi.Color.Red ++ fansi.Bold.On
+    case DiagnosticKind.Warning => fansi.Color.Yellow ++ fansi.Bold.On
+  }
+
+  private val subject: String = diagnosticKind match {
```

Comment:
`val` create a field in the instance, doesn't it? Why do we need field `subject` when there already is a field `diagnosticKind` which is also redundant as the kind is derived based on type `diagnostics`.

These fields should be turned into methods unless there is some reason why it is better to trade some negliable CPU gain for memory.